### ISSUE TITLE
Messages to nine.testrun.org are rejected with valid SPF

### ISSUE THREAD

**Issue by ell1e** [deltachat/chatmail#416]
Messages to nine.testrun.org appear to be rejected even with valid SPF:

```
Final-Recipient: rfc822; [REDACTED@nine.testrun.org](mailto:REDACTED@nine.testrun.org)
Original-Recipient: [rfc822;REDACTED@nine.testrun.org](mailto:rfc822;REDACTED@nine.testrun.org)
Action: failed
Status: 5.7.1
Remote-MTA: dns; nine.testrun.org
Diagnostic-Code: smtp; 554 5.7.1 No valid DKIM signature found
```

This seems like a problem, because from my experience no major provider including even the occasionally picky GMail require DKIM but only *either* DKIM or SPF as far as I know. Therefore, people will be unable to use DeltaChat when normally, their e-mail would work fine in that situation. Since SPF seems like it would likely suffice unless there's a man-in-the-middle at the ISP level, at which point PGP seems like the better defense than DKIM, I'm personally also confused why this requirement has much benefit on a technical level (but don't trust me on this, just my two cents).

- Operating System (Linux/Mac/Windows/iOS/Android):
- Delta Chat Version: Desktop v1.46.8 Core v1.142.12 SQLite 3.45.3 electron 30.3.1 node 20.15.1
- Expected behavior: e-mail is accepted when SPF is valid, even when there's no DKIM.
- Actual behavior: e-mails are rejected even with valid SPF
- Steps to reproduce the problem: use an e-mail server that uses SPF but not DKIM
- Screenshots: 
![Screenshot_20241001_185610](https://github.com/user-attachments/assets/45f7a436-f17c-4655-806d-193944c881ed)

- Logs:

**Comment by missytake** [comment]
First of all, that DKIM signatures are required is the expected behavior for chatmail servers (like nine.testrun.org), see https://github.com/deltachat/chatmail?tab=readme-ov-file#email-authentication

I don't remember 100% what the arguments were, just that your arguments were considered as well. In general, the discussion was centered around "let's make From forgery as hard as possible, even if it breaks compatibility with some classic email servers". The priority use case of chatmail servers is chatting, not email; and users expect from chatting that they can rely on the other side being who they claim to be, and senders can not simply forge the `From` header. Whether this is enforced or not is not dependent on DKIM of course, not even on SPF or DMARC; but it's the best indicator we have, and it's a necessary premise that at least the domain part of the `From` header is correct, even if the localpart/username is not enforced to belong to an authenticated account.

The email ecosystem has tried to fight From forgery for decades with SPF, DKIM, and DMARC, to limited success. The result is that many DMARC policies have `p=none`, SPF softfail policies, and don't really give us the guarantee we'd like to see. DKIM is the mechanism which is closest to what we want; if there is a valid DKIM signature, we can be sure that the server which sent is allowed to send emails for this domain.

(In the beginning we experimented with other policies around DMARC, but we quickly realized that if DKIM is correct we can ignore what SPF says, and if DKIM is missing we can not rely on SPF. The only thing DKIM was missing that DMARC provided was the `adkim` parameter which checks whether the `From` domain == the DKIM signature, and we simply require this in an additional step, instead of relying on DMARC, which can also specify to ignore whether `From` domain and DKIM signature match.)

Yes, this breaks interoperability with some other email servers, but then again this interoperability is already limited, as chatmail accounts can not send out unencrypted emails for example. Compared to that, DKIM is set up automatically for chatmail servers, and usually just works out of the box. Our hope is that this motivates admins to add proper DKIM setups; we can allow ourselves to make this hard requirement, because 100% interoperability with email is not expected from chatmail anyway.

Adding DKIM is not super hard, and I can recommend https://www.learndmarc.com/ as a test site which does a great job at both explaining the concepts around DMARC and helping admins to configure their server correctly. Then again I have to admit I also co-manage a server which does not DKIM-sign^^ If you manage the server in the issue, and need help with setting it up, drop me an email at <missytake@systemli.org> :) it might take some days for me to respond, but I'd be happy to help.

**Comment by ell1e** [comment]
I do manage the mail server, but I still don't quite see the point. Why not use SPF hard fail and PGP instead? This seems to be the standard by now. I didn't setup DKIM because every extra option risks maintenance trouble down the line, and it's simply not required in practice for any normal use of e-mail that I've encountered so far (while SPF **is**). Based on this, I think making DKIM a hard fail is a problem, unless your plan is to have two separate worlds of people using chatmail that can chat to each other, and the others "outside", with no interoperability.

**Comment by ell1e** [comment]
Nothing you do seems like it would work well to assume safely the OpenPGP key is actually a valid one for that user. If you wanted to improve that, I think it might be a good idea to show the key id here:
![Screenshot_20241001_203234](https://github.com/user-attachments/assets/5f3ad276-4804-41f4-9804-a201b296ee89)
And to allow accessing a recently used OpenPGP list for a user here:
![Screenshot_20241001_213112](https://github.com/user-attachments/assets/2121478b-0959-43bc-ba0c-faece121cc24)
And to show a big visible warning if the OpenPGP key for a message from a user differs from the key used for the previous message from that user.

As for SPF, you could if you wanted just check the SPF entry for the FROM address host as well in case DKIM isn't available. I realize that's not the spec, but you'd only be allowing more mails to pass through, not less.

The complexity argument you mention sadly also applies on the sender side, e.g. DKIM is very complex to set up for me and in my opinion doesn't help with making OpenPGP much better, so I don't bother. I can't imagine I'm the only one.

**Comment by ell1e** [comment]
For what it's worth, I'm not really asking that you check SPF but rather that you don't enforce DKIM. However, I realize in practice that might be the same, since it seems like many big providers want either SPF or DKIM to be present and working which seems like a reasonable baseline.

**Comment by ell1e** [comment]
Why not allow seeing that information with an advanced mode? And the info on key change sounds good, although I wonder if it also should be a warning unless the new key is signed by the old key. Does Delta Chat check that?

You may also want to consider something like threema, where any new user defaults to some "Yellow"/medium trust level and has a short fingerprint shown (which would be based on the OpenPGP key) and then needs to be manually marked "Green"/high trust by the user if marked as trusted. And whenever the key changes, it could go to "Red"/low trust.

All of those approaches sound more useful than DKIM blocks: I believe not even matrix which I would think is in many ways is more paranoid than Delta Chat is, downright rejects messages with invalid signatures or invalid sources. But this is what chatmail is effectively doing if no DKIM means block, rather than e.g. showing a warning but letting the message pass. I get the desire to verify the sender though, so I understand it on some level.

**Comment by ell1e** [comment]
Ah, very cool, thanks for the info! That's super useful, I like to verify and compare the keys in real life sometimes. Although I realize it'll go past most average users, which I agree is an issue.

**Comment by ell1e** [comment]
I did just make a UI proposal for better verification here, for what it's worth: https://support.delta.chat/t/feature-suggestion-manual-fingerprint-verification-securejoin-alternative-for-desktop-users/3276

I think focusing purely on QR codes isn't a good idea, since some users don't use devices that can handle these. DKIM doesn't really solve that either, but it seems relevant to this discussion of general sender trust.

**Comment by ell1e** [comment]
Yeah that makes sense. I'll see if I can get DKIM to work, for what it's worth. But I'm hoping SPF as an alternative will at least remain a consideration, if only for access for others with their own email. The other person I know that runs their own email also hasn't DKIM enabled, so it seems like it's not the most popular among self-hosters.

**Comment by ell1e** [comment]
It's not necessarily a bad idea. But since client can use whatever e-mail address and "normal" e-mail addresses usually don't require DKIM on *all* incoming mail since that's not practical, Delta Chat probably shouldn't be designed to rely on this behavior of the SMTP server. But then I wonder why have it ... As somebody suggested on the forum, one way to make this more meaningful and consistent would be if Delta Chat was changed to actually check DKIM inside the chat client.

**Comment by ell1e** [comment]
After some poking around more, I think rejecting messages without DKIM is actually possibly a good idea. However, I think Delta Chat should also show DKIM status on the SMTP server's `Authentication-Results` header, including both if DKIM failed or is missing.
